core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -13,22 +13,55 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
+import java.util.function.Supplier;
 
 public class ThreadContextClassLoader
-        implements Closeable
 {
-    private final ClassLoader originalThreadContextClassLoader;
+    private ThreadContextClassLoader() {}
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
+    public static <T> T withClassLoader(ClassLoader newThreadContextClassLoader, Supplier<T> supplier)
     {
-        this.originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
         Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
     }
 
-    @Override
-    public void close()
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(newThreadContextClassLoader, (Supplier<T>) supplier);
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
+            extends Supplier<T>
```

Comment:
This won't work for:

> as Confluent's SchemaRegistryClient has throws IOException, RestClientException pretty much on every method.

since the two exception types are not in direct hierarchy

we would need to accept E1, E2... would that work?

Reply:
If this doesn't work - then `ClassLoaderSafeSchemaRegistryClient` is not covered by any functional tests in the CI :) 

The way I understand it, `... throws E1 extends Exception, E2 extends Exception` is reduced to `... throws Exception` which kind of makes sense as each E1 and E2 'is an' Exception. I believe the compiler interprets it in a similar way:

[JLS 18.2.5 Checked Exception Constraints](https://docs.oracle.com/javase/specs/jls/se11/html/jls-18.html#jls-18.2.5)

>  Otherwise, let E1, ..., En be the types in the function type's throws clause that are not proper types. If the lambda expression is implicitly typed, let its parameter types be the function type's parameter types. If the lambda body is a poly expression or a block containing a poly result expression, let the targeted return type be the function type's return type. Let X1, ..., Xm be the checked exception types that the lambda body can throw (§11.2). Then there are two cases:

> If n = 0 (the function type's throws clause consists only of proper types), then if there exists some i (1 ≤ i ≤ m) such that Xi is not a subtype of any proper type in the throws clause, the constraint reduces to false; otherwise, the constraint reduces to true.

> If n > 0, the constraint reduces to a set of subtyping constraints: for all i (1 ≤ i ≤ m), if Xi is not a subtype of any proper type in the throws clause, then the constraints include, for all j (1 ≤ j ≤ n), ‹Xi <: Ej›. In addition, for all j (1 ≤ j ≤ n), the constraint reduces to the bound throws Ej.

Reply:
`ThrowingSupplier` does not extend `Supplier` since https://github.com/trinodb/trino/pull/8507/commits/e60dfc9d2b3e1fc80f1d3e4520670f890c8acf1d, resolving this thread.